Migrate sympy.physics.quantum to use Kinds and _constructor_postprocessor_mapping

ellisonbg opened the pull request:
The goal of this effort is to remove the special methods like `__mul__`, `__rmul__`, and `__pow__` from the quantum classes. Instead, we are moving to using the new `Kind` and `_constructor_postprocessor_mapping` logic.

Before review/merge:

- [x] Implement kinds for bras, kets and operators.
- [x] Replace custom `__mul__` and `__pow__` method with `_constructor_postprocessor_mapping` based transformed.
- [x] Add tests.
- [x] Make sure all tests and doctests pass.
- [x] Add docstrings.
- [x] Add note to release notes about `tensor_product_simp` and friends being deprecated.

Follow up issues to open:

- [x] Open issue documenting issues with kind and `_constructor_postprocessor_mapping` APIs.
- [x] Open issue about introducing a binary operator for tensor products.
- [x] Open issue for proper kind handling of `Operator*Ket` and `Bra*Operator` (improvements to the kind dispatcher are needed first.

#### References to other Issues or PRs

Fixes #27248
Fixes #19540 
Fixes #19538 

#### Brief description of what is fixed or changed

* Add `OperatorKind`, `BraKind`, `KetKind`.
* Add logic that uses `_constructor_postprocessor_mapping` to transform quantum expressions when building `Mul` and `Pow` expressions.
* Use multiple-dipatch to define rules for the transformations.
* Remove all `__mul__` and `__pow__` methods from quantum package.
* All combinations of bras and kets are now consistently identified as inner/outer products, with inner products taking taking the highest priority when there is ambiguity.
* TypeError is raised when trying to build tensor products using `*`, such as `Ket()*Ket()` or `Bra()*Bra()`.
* Breaking change: `tensor_product_simp` and its helper functions have been deprecated as the transformations they apply are now handled automatically using `_constructor_postprocessor_mapping`.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* physics.quantum
  * Removed custom `__mul__` and `__pow__` methods from quantum classes. This logic has been replaced by `_constructor_postprocessor_mapping` based transforms that are more complete and consistent.
  * Created kinds for bras, kets and operators, with kind dispatch resolution.
  * All combinations of bras and kets are now consistently identified as inner/outer products, with inner products taking taking the highest priority when there is ambiguity.
  * TypeError is raised when trying to build tensor products using `*`, such as `Ket()*Ket()` or `Bra()*Bra()`.
  * Breaking change: `tensor_product_simp` and its helper functions have been deprecated as the transformations they apply are now handled automatically using `_constructor_postprocessor_mapping`.


<!-- END RELEASE NOTES -->

ellisonbg committed:
quantum: Initial work on adding kinds and using transforms for Mul and Pow.

The goal of this effort is to remove the special methods like __mul__, __rmul__, and __pow__ from the quantum classes. Instead, we are moving to using the new Kind and _constructor_postprocessor_mapping logic. This commit has the initial pieces in place, but more work is required.

ellisonbg committed:
quantum: Continued work on the kind and Mul/Pow transformations.

This commits cleans things up in the PR and gets the tests working again.

ellisonbg committed:
quantum: Fixes in qapply

The qapply function had some bugs related to Sums that needed fixing. Some of these happened during a rebase, others were in master.

ellisonbg committed:
quantum: Rename statekind.py to kind.py

ellisonbg committed:
quantum: Consolidating quantum kinds into one file (kind.py)

ellisonbg committed:
quantum: Adding tests for kinds in quantum

This commit adds a relatively complete set of tests for kinds in the quantum module. This also continues to refactor the kind logic in quantum into a single place (kind.py).

ellisonbg committed:
quantum: Fixing tests in operator and sho1d

In the work of this PR a few tests were broken. This commit fixes the tests.

ellisonbg committed:
quantum: Remove commented out __mul__

A __mul__ was commented out and I forgot to delete it. This commits fixes that.

ellisonbg committed:
quantum: Clean-up of code for PR

Cleaning up a few locations in the code, removing duplicate imports, fixing indentation.

ellisonbg committed:
quantum: Adding back missing import

Last commit removed a needed import. Adding that back.

ellisonbg committed:
quantum: More cleanup and move transforms to a new file.

The transforms that are registered with Mul and Pow as postprocessors were in state.py but are now in their own modue (transforms.py).

ellisonbg committed:
quantum: Minor fixes to make Ruff happy.

I had a bunch of unused imports and other minor issues that Ruff was complaining about. This fixes those issues.

ellisonbg committed:
quantum: Fixing more imports.

More import errors from moving the transforms to their own file. Should be fixed now.

ellisonbg committed:
quantumm: Fixing doctests

A number of the doctests needed to be updated because of the improved logic introduced in this PR.

ellisonbg committed:
quantum: Fixing flake8 unused import in docstring

Because of docstring refactoring there was an unused import of Operator in a docstring that this commit removes.

ellisonbg committed:
quantum: Finish tests for quantum transforms.

This commits adds tests for all of the cases in transforms.py. I also needed for fix some existing tests in test_spin.py because the form of the expressions are different (more consistent in this PR).

ellisonbg committed:
quantum: Fixes for ruff check

Remove unused imports in test_transforms.py.

ellisonbg committed:
quantum: Improve tensor product handling

Previously this PR tried to combine products of kets or bras into tensor products using normal multiplication. This was leading to a number of confusing situation where regular* multiplication could mean different things in different situations (in quantum expressions). This has been resolved by transforms now raising TypeError if you try to combine bras and kets this way. Instead, users need to explicitly use TensorProduct.

In the future, I would like to introduce a new binary operator that is a shorthand for TensorProduct, but that is work for later.

Another side effect of this work is that the tensor_product_simp functions and their helpers can be deprecated, which this commit also does.

ellisonbg committed:
quantum: Add docstrings and inline comments to PR

In this commit I have added docstrings and inline comments for (hopefully) everything that needs it.

ellisonbg commented:
OK, this should be ready for review (assuming I haven't broken anything while adding docstrings).

oscarbenjamin commented on sympy/physics/quantum/kind.py:
```diff
@@ -0,0 +1,103 @@
+# Kind resolution.
+#-----------------------------------------------------------------------------
+
+# Note: We can't currently add kind dispatchers for the following combinations
+#       as the Mul._kind_dispatcher is set to commutative and will also
+#       register the opposite order, which isn't correct for these pairs:
+#
+# 1. (_OperatorKind, _KetKind)
+# 2. (_BraKind, _OperatorKind)
+# 3. (_BraKind, _KetKind)
```
Maybe we could change this so that `Mul._kind_dispatcher` handles commutativity. I'm not sure if it assumes commutativity because that is more efficient somehow.

This is one of those situations where it would be better if there was a non-commutative NCMul like in other CAS.

ellisonbg replied:
Yeah, the dispatcher takes a `commutative` argument in the constructor, but not in the `add` method. I believe it will be straightforward to add the argument in the `add` method as an override and will open an issue/PR about this and follow up with fixes in the quantum kind module.

oscarbenjamin commented on sympy/physics/quantum/qapply.py:
```diff
@@ -68,18 +80,22 @@ def qapply(e, **options):
-    if e == 0:
-        return S.Zero
+    if isinstance(e, (int, complex, float)):
+        e = sympify(e)
```
Why not just call `sympify` unconditionally? Actually `_sympify` is better because it doesn't parse strings.

ellisonbg replied:
Will fix this now.

oscarbenjamin commented on sympy/physics/quantum/tests/test_transforms.py:
```diff
@@ -0,0 +1,75 @@
+    # In particular, we are not trying to interpret regular ``*`` multiplication
+    # as a tensor product.
+    with raises(TypeError):
+        assert k1*k1 == TensorProduct(k1, k1)
```
Which part of this is raising TypeError? Is it the `==` or the `k1*k2`. The with block should only have the minimal code that raises the error.

ellisonbg replied:
Yeah, jusst the `k1*k2` part, will fix that now.

oscarbenjamin commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,10 +76,28 @@ SymPy deprecation warnings.
+``sympy.physics.quantum.transforms`` module.
+
+If you are using these functions in your code, you can remove them as
+they are now reduntant.
```
Is it likely that downstream library code uses these functions?

If so I think it is unfriendly to start emitting a deprecation warning immediately. The functions can continue to work even if they are not needed any more and then when the new functionality has been available in a few releases it is a better time to add the warning.

ellisonbg replied:
The old implementations of these functions are both slower, duplicative, and not as correct/complete as the automatics transforms. I am a bit hesitant to change the functions to a pass through without telling users. At the same time, leaving the original implementation and not warning leaves them using a function that isn't needed and doesn't work as well as they might like. Thoughts on which way to go?

ellisonbg replied:
Also, I don't know that these are heavily used in downstream library code as they don't fully work as one would want.

oscarbenjamin replied:
If it is unlikely that the functions are being used anywhere then it is fine.

In general though some downstream libraries try to support multiple versions of sympy. What that means is that if we add a new thing and deprecate an old thing then I think there needs to be some releases in between so that downstream code can start using the new thing as soon as they see the deprecation warning rather than needing to do something like:
```python
if sympy.__version__ < 1.14:
   from sympy import ....
else:
   ...
```

ellisonbg committed:
quantum: Simplify with raises tests

The with raises tests now include the minimal code required to raise the exception.

ellisonbg committed:
quantum: Improve sympify logic in qapply

Using _sympify rather than sympify in places. Stop type checking for parts of it.

ellisonbg commented:
Review points addressed.

oscarbenjamin commented on sympy/physics/quantum/qapply.py:
```diff
@@ -28,6 +29,17 @@
 #-----------------------------------------------------------------------------
 
+
+def ip_doit_func(e):
+    """Transform the inner products in an expression by calling ``.doit()``."""
+    return e.replace(InnerProduct, lambda *args: InnerProduct(*args).doit())
+
+
+def sum_doit_func(e):
+    """Transform the sums in an expression by calling ``.doit()``."""
+    return e.replace(Sum, lambda *args: Sum(*args).doit())
```
There should really be a general way of doing this. Pretty much all library code that calls `.doit()` really should only call `.doit()` on some particular type like this.

oscarbenjamin commented:
I haven't checked anything for mathematical correctness but I will trust you on that.

This looks good to me so merging.
